src/input/pick_color_grab.rs

```diff
@@ -0,0 +1,223 @@
+use smithay::backend::allocator::Fourcc;
+use smithay::backend::input::ButtonState;
+use smithay::backend::renderer::element::utils::{Relocate, RelocateRenderElement};
+use smithay::backend::renderer::gles::GlesRenderer;
+use smithay::input::pointer::{
+    AxisFrame, ButtonEvent, CursorImageStatus, GestureHoldBeginEvent, GestureHoldEndEvent,
+    GesturePinchBeginEvent, GesturePinchEndEvent, GesturePinchUpdateEvent, GestureSwipeBeginEvent,
+    GestureSwipeEndEvent, GestureSwipeUpdateEvent, GrabStartData as PointerGrabStartData,
+    MotionEvent, PointerGrab, PointerInnerHandle, RelativeMotionEvent,
+};
+use smithay::input::SeatHandler;
+use smithay::utils::{Logical, Physical, Point, Scale};
+
+use crate::niri::State;
+use crate::render_helpers::render_to_vec;
+
+pub struct PickColorGrab {
+    start_data: PointerGrabStartData<State>,
+    temporary_location: Option<Point<f64, Logical>>,
+}
+
+impl PickColorGrab {
+    pub fn new(start_data: PointerGrabStartData<State>) -> Self {
+        Self {
+            start_data,
+            temporary_location: None,
+        }
+    }
+
+    fn on_ungrab(&mut self, state: &mut State) {
+        if let Some(tx) = state.niri.pick_color.take() {
+            let _ = tx.send_blocking(None);
+        }
+        state
+            .niri
+            .cursor_manager
+            .set_cursor_image(CursorImageStatus::default_named());
+        state.niri.queue_redraw_all();
+    }
+
+    fn pick_color_at_point(location: Point<f64, Logical>, data: &mut State) -> Option<[u8; 4]> {
+        let (output, pos_within_output) = data.niri.output_under(location)?;
+
+        data.backend
+            .with_primary_renderer(|renderer| {
+                let scale = Scale::from(output.current_scale().fractional_scale());
+                let physical_pos: Point<i32, Physical> =
+                    pos_within_output.to_physical_precise_round(scale);
```

Comment:
This might need to be a `floor` with a FIXME comment currently because I suspect that at the right/bottom edge of monitors it will return one past.